### PR TITLE
Add support for Binlog Transaction Compression

### PR DESCRIPTION
**Pull request by altmannmarcelo**
This patch adds support for Binlog Transaction Compression, which has been available since MySQL 8.0.20. It is disabled by default, but it is widely used in production.

The patch adds a new event type, TransactionPayloadEvent and its necessary methods to parse binlog event TRANSACTION_PAYLOAD_EVENT(0x28).

The patch also adds a method to decompress the payload if it has been compressed with ZSTD.

More information about the feature can be found here:

https://dev.mysql.com/worklog/task/?id=3549
https://github.com/mysql/mysql-server/commit/1e5ae348c6f760b16a3972d84956358606a30041

### CODE REVIEW

**Comment by altmannmarcelo** [comment]
@blackbeam I have a GDocs document that helps to understand the new binlog event format, such as manually decoding it - https://docs.google.com/document/d/1Ypnsua2W-MQfGoogPXw2u_5y3QW3qdQeXkzTQyKtEPM/edit?usp=sharing

I wrote a small test program to validate the new format:

```rust
use mysql_async::prelude::Queryable;
use mysql_common as mysql;
use mysql_common::binlog::consts::EventType;
use mysql_common::binlog::{self, events};

#[tokio::main]
async fn main() -> Result<(), mysql_async::Error> {
    let mut reader = binlog::EventStreamReader::new(binlog::consts::BinlogVersion::Version4);
    // Payload event has checksum disabled. Use different reader to read it.
    let mut reader_no_checksum =
        binlog::EventStreamReader::new(binlog::consts::BinlogVersion::Version4);
    let database_url = mysql_async::Opts::from_url("mysql://root:@localhost:3310/test")?;

    let pool = mysql_async::Pool::new(database_url);
    let mut conn = pool.get_conn().await?;

    conn.query_drop("CREATE TABLE IF NOT EXISTS tb1(ID INT); FLUSH BINARY LOGS;")
        .await?;
    let query = "SHOW MASTER STATUS";
    let pos: mysql::Row = conn
        .query_first(query)
        .await?
        .ok_or_else(|| mysql_async::Error::Driver(mysql_async::DriverError::ConnectionClosed))?;
    let file: String = pos.get(0).expect("Binlog file");
    let offset: u64 = pos.get(1).expect("Binlog offset");

    conn.query_drop("SET binlog_transaction_compression=ON; INSERT INTO tb1 VALUES (1)")
        .await?;
    let cmd = mysql_common::packets::ComRegisterSlave::new(2);
    conn.query_drop("SET @master_binlog_checksum='CRC32'")
        .await?;
    conn.write_command(&cmd).await?;
    conn.read_packet().await?;

    let cmd = mysql_common::packets::ComBinlogDump::new(2)
        .with_filename(file.as_bytes())
        .with_pos(offset.try_into().expect("Impossible pos"));
    conn.write_command(&cmd).await?;
    conn.read_packet().await?;

    let mut keep = true;
    while keep {
        let packet = conn.read_packet().await?;
        assert_eq!(packet.first(), Some(&0));
        let binlog_event = reader.read(&packet[1..])?;
        match binlog_event.header().event_type().map_err(|ev| {
            println!("Unknown event type: {:?}", ev);
            mysql_async::Error::Driver(mysql_async::DriverError::ConnectionClosed)
        })? {
            EventType::TRANSACTION_PAYLOAD_EVENT => {
                let ev: events::TransactionPayloadEvent = binlog_event.read_event()?;
                println!("Received TRANSACTION_PAYLOAD_EVENT - Compressed Size: {}, Uncompressed Size: {}", ev.payload_size(), ev.uncompressed_size());
                let buff = ev.payload_raw_decompressed();
                let mut read_pos = 0;
                while read_pos < buff.len() {
                    let binlog_ev = reader_no_checksum.read(&buff[read_pos..])?;
                    read_pos += binlog_ev.header().event_size() as usize;
                    println!(
                        "Received decompressed event: {:?}",
                        binlog_ev.header().event_type()
                    );
                }
                keep = false;
            }
            ev => {
                println!("Received other event type: {:?}", ev);
            }
        }
    }
    // Dropped connection will go to the pool
    drop(conn);

    // The Pool must be disconnected explicitly because
    // it's an asynchronous operation.
    pool.disconnect().await?;

    // the async fn returns Result, so
    Ok(())
}

``` 

Example output:
```
$ ./target/debug/binlog 
Received other event type: FORMAT_DESCRIPTION_EVENT
Received other event type: ANONYMOUS_GTID_EVENT
Received TRANSACTION_PAYLOAD_EVENT - Compressed Size: 129, Uncompressed Size: 179
Received decompressed event: Ok(QUERY_EVENT)
Received decompressed event: Ok(TABLE_MAP_EVENT)
Received decompressed event: Ok(WRITE_ROWS_EVENT)
Received decompressed event: Ok(XID_EVENT)
```

File: src/binlog/events/transaction_payload_event.rs [hunk 1]
```diff
@@ -0,0 +1,296 @@
...
+                            Err(io::Error::new(
+                                io::ErrorKind::InvalidData,
+                                format!("Unknown field type: {:?}", val),
+                            ))?;
+                        }
+                    }
+                }
+                Err(val) => {
+                    let length = buf.eat_lenenc_int();
+                    buf.skip(length as usize);
```
Comment (blackbeam) [inline src/binlog/events/transaction_payload_event.rs:262]: net_field_length format error seem to be fatal, and there is nothing to skip here.
I believe we'd better to replace this `match buf.read_lenenc_int() { .. }` with just `let val = buf.read_lenenc_int()?`.

File: src/binlog/events/transaction_payload_event.rs [hunk 2]
```diff
@@ -0,0 +1,296 @@
...
+            };
+        }
+
+        Ok(ob)
+    }
+}
+
+impl MySerialize for TransactionPayloadEvent<'_> {
+    fn serialize(&self, buf: &mut Vec<u8>) {
+        self.payload_size.serialize(&mut *buf);
```
Comment (blackbeam) [inline src/binlog/events/transaction_payload_event.rs:277]: Looks like it must follow TLV serialization (see [here](https://github.com/mysql/mysql-server/commit/1e5ae348c6f760b16a3972d84956358606a30041#diff-184e9a7d8a58f080974e475d4199fe5c6da5518c8a2811cc5df5988c8f9e9797R158-R185)). The END_MARK must also be here.

File: src/binlog/events/transaction_payload_event.rs [hunk 3]
```diff
@@ -0,0 +1,296 @@
...
+    pub fn payload_raw(&'a self) -> &'a [u8] {
+        self.payload.as_bytes()
+    }
+
+    /// Returns raw payload decompressed
+    pub fn payload_raw_decompressed(self) -> Vec<u8> {
+        if self.algorithm == TransactionPayloadCompressionType::NONE {
+            return self.payload_raw().to_vec();
+        }
+        let decoded = zstd::stream::decode_all(self.payload.as_bytes()).unwrap();
```
Comment (blackbeam) [inline src/binlog/events/transaction_payload_event.rs:100]: Can we use `decompressed_size` to preallocate a buffer for `zstd::stream::copy_decode` here?

**Review by blackbeam — CHANGES_REQUESTED** [review]
Thanks for PR!

Could you please create a binlog file to incorporate this into the [test suite](https://github.com/blackbeam/rust_mysql_common/blob/7b28d6c4539abac6c5bb617bc71577bc148102ea/src/binlog/mod.rs#L634)? We'd better to cover both `CompressionType::NONE` and `CompressionType::ZSTD`.

File: src/binlog/events/transaction_payload_event.rs [hunk 4]
```diff
@@ -0,0 +1,296 @@
...
+                                ))?
+                                }
+                            };
+                            ob.uncompressed_size = RawInt::new(val);
+                            continue;
+                        }
+                        Err(val) => {
+                            let length = buf.eat_lenenc_int();
+                            buf.skip(length as usize);
+                            Err(io::Error::new(
```
[the remark this answers is not in the document]
Comment (altmannmarcelo) [reply src/binlog/events/transaction_payload_event.rs]: Done

File: src/binlog/events/transaction_payload_event.rs (hunk 1 above)
Comment (altmannmarcelo) [reply src/binlog/events/transaction_payload_event.rs:195]: Done

File: src/binlog/events/transaction_payload_event.rs (hunk 2 above)
Comment (altmannmarcelo) [reply src/binlog/events/transaction_payload_event.rs]: Indeed. Done

File: src/binlog/events/transaction_payload_event.rs (hunk 3 above)
Comment (altmannmarcelo) [reply src/binlog/events/transaction_payload_event.rs]: Done

**Comment by altmannmarcelo** [comment]
@blackbeam 

Having a `CompressionType::NONE` might be a bit tricky. From what I remember, ZSTD will write uncompressed frames if it does not see enough range redundancies and write the frame uncompressed. 

I have added a binlog with compressed events and extended the test to validate it can decompress the Payload Event and read proper events in the decompressed buffer.

**Pull request merged**